### ISSUE TITLE
Error prone vector interfaces due to reference semantics

### ISSUE THREAD

**Issue by jamiebramwell** [mfem/mfem#3531]
I recently came across a bug where the underlying `data` pointer in an `mfem::Vector` was changed in the middle of a Sundials nonlinear function callback. This caused the connection between the `SundialsNVector` and the Sundials-controlled `N_Vector` to be broken. While the code ran just fine, the solver failed to converge and resulted in a difficult to trace error.

In this particular case it was due to the use of a move assignment operator, but this error could also be triggered by `NewMemoryAndSize` or `MakeRef` commands in the middle of a callback function. Any time where reference semantics are implicitly required by the use case of `mfem::Vector` could have similar problems.

A few possible solutions:
1. Create a `VectorView` or `VectorRef` class that explicitly denotes these const memory address but mutable data value scenarios.
2. Try to deduce this case from ownership flags and detect when these contracts are broken (see https://github.com/mfem/mfem/pull/3530 for an example).

I like 1 mfe, but it would likely require a lot of interface changes. Any thoughts?

**Comment by samuelpmishLLNL** [comment]
do you have a minimal reproducer to look at?

**Comment by v-dobrev** [comment]
In a way this can be seen as a deficiency of C++: it supports only two levels of access: non-const and const. In the case of `Vector`, it would've been great if we could add another (custom?) access level (say `const_alloc`) where we can modify the entries of the `Vector` but we cannot change the memory allocation it uses.

Of course, this is not going to be possible any time soon, so we need to use another approach. Using a separate class `VectorView`, `VectorRef`, or something like that seems a good way to go.

**Comment by samuelpmishLLNL** [comment]
> In a way this can be seen as a deficiency of C++: it supports only two levels of access: non-const and const.

This is not true, there are different kinds of `const`-ness for pointers: `const` pointers and pointers-to-`const`.

> it would've been great if we could add another (custom?) access level (say const_alloc) where we can modify the entries of the Vector but we cannot change the memory allocation it uses.

What you're describing here is the behavior of a `const` pointer.

**Comment by v-dobrev** [comment]
What I was trying to say is that C++ does not allow you to have a third access mode in the same class -- it has non-`const` and `const` modes only. Any other semantic has to be done using a different class.

**Comment by samuelpmishLLNL** [comment]
I think the more sinister form of the reproducer is something like:

```cpp
  inner_sundials_nvector = mfem::Vector(5);
```

where, it's not immediately obvious that a move is happening at all. 

Depending on the semantics of the container, `operator=` could be a deep copy, shallow copy, move. `mfem::Vector` tries to support all of those different kind of semantics in a single container, which makes it difficult to reason about.

**Comment by v-dobrev** [comment]
I think we are missing some important context here: I think the unexpected behavior begins in the following method: https://github.com/mfem/mfem/blob/888b8eca6dfe98e4838c95b4a4fbddd0a51141cb/linalg/sundials.cpp#L1809-L1820
In this implementation, it is assumed that the call `self->oper->Mult(mfem_u, mfem_fu)` is not going to change the memory to which `mfem_fu` points to. However, the method signature of `Operator::Mult`, https://github.com/mfem/mfem/blob/888b8eca6dfe98e4838c95b4a4fbddd0a51141cb/linalg/operator.hpp#L88-L89
does not forbid implementations of `Mult` to do just that -- e.g. to construct a new `Vector` and move-assign it to the second (result) argument of `Mult`. And that's exactly what was happening in a user-defined `Mult` in a user-defined sub-class of `Operator`.

With this context in mind, I don't think we can blame the `Vector` class implementation here. I'd say the problem is in the lack of expressiveness in the `Mult` call that is used. The lacking information here (for both the compiler and for users) is that `Mult` should not change the memory used by second argument and instead just modify the entries of the `Vector`. One way to express this is to introduce a new `Mult` method in `Operator` that uses second argument (i.e. result type) with a type different from `Vector &` that behaves as a "Vector" for the most part but does not allow you to change its memory allocation, e.g. `[const] VectorView &`, or `[const] VectorRef &`.

**Comment by v-dobrev** [comment]
Another approach to help users detect this kind of issue at runtime is to add checks after calls of `Mult` that expect the result vector to not be re-located. The check will be to make sure the `Vector` was not re-located and trigger an error if that happened. Alternatively, instead of triggering an error, the implementation can just copy the data to the memory location where it was supposed to be.
